### PR TITLE
Optimal coefficients, fixes #8847

### LINKED ISSUE #8847: `machine_learning/linear_regression.py` doesn't give optimal coefficients
### Feature description

Related to issue #8844

TL;DR: the current implementation doesn't give optimal solutions, the current implementation calculates SSE wrong, and we should add an implementation of a numerical methods algorithm that actually gives optimal solutions

In `machine_learning/linear_regression.py`, add the following code at the bottom of the `main()` function:
```py
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

data = np.asarray(data.astype(float))
X = data[:, 0].reshape(-1, 1)
y = data[:, 1]

reg = LinearRegression().fit(X, y)
print(f"Sklearn coefficients: {reg.intercept_}, {reg.coef_}")

sse = np.sum(np.square(reg.predict(X) - y.T))
print(f"{sse = }")
print(f"mse = {sse / len(y)}")
print(f"half mse = {sse / (2 * len(y))}")

plt.scatter(X, y, color="lightgray")
plt.axline(xy1=(0, theta[0, 0]), slope=theta[0, 1], color="red", label="Gradient descent")
plt.axline(xy1=(0, reg.intercept_), slope=reg.coef_, color="blue", label="Sklearn")
plt.legend()
plt.show()
```
This code performs ordinary least squares (OLS) linear regression using `sklearn` as a point of reference to compare the current implementation against. It then calculates the sum of squared errors (SSE), mean squared error (MSE), and half of the MSE. To compare the outputs visually, the code uses `matplotlib` to plot the `sklearn` regression line and the regression line produced by the current implementation.

The code produces the following command line output:
```
...
At Iteration 100000 - Error is 128.03882
Resultant Feature vector: 
-9.34325
1.53067
Sklearn coefficients: -15.547901662158367, [1.6076036]
sse = 253795.17406773588
mse = 253.79517406773587
half mse = 126.89758703386794
```
As we can see, what the implementation calculates as the SSE (128.03882) is actually half of the MSE, meaning that the `sum_of_square_error` function is incorrect and needs to be fixed. Why the implementation is using _half_ of the MSE, I have no clue.

Furthermore, we can see that both the regression coefficients and the errors are slightly off. This is because the current implementation works via gradient descent, meaning that it can only _approximate_ the OLS regression coefficients. Meanwhile, libraries like `numpy` and `sklearn` calculate the mathematically optimal coefficients using [numerical methods](https://en.wikipedia.org/wiki/Numerical_methods_for_linear_least_squares).
linear_regression_plot
Although using gradient descent to perform linear regression does work, it's still suboptimal and (AFAIK) it's not how linear regression is actually performed in practice. We can still include this implementation, but we should definitely also include an implementation of an optimal numerical method.

### PR DESCRIPTION
**Pull request by liakdimi1**
### Describe your change:
FIxes #8847
I found the optimal iterations and learning rate for achieving the closest coefficients
Στιγμιότυπο οθόνης (266)
* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: machine_learning/linear_regression.py [hunk 1]
```diff
@@ -70,8 +69,8 @@ def run_linear_regression(data_x, data_y):
     :param data_y  : contains the output (result vector)
     :return        : feature for line of best fit (Feature vector)
     """
-    iterations = 100000
-    alpha = 0.0001550
+    iterations = 250000
+    alpha = 0.000326
```
Comment (tianyizheng02) [inline machine_learning/linear_regression.py:73]: This won't give the optimal coefficients. Like I said in my original issue, gradient descent is an approximation algorithm _by definition,_ so tweaking the parameters will only give you a better approximation. My point with the original issue is that we should add a _different_ algorithm, one that is a _direct method_ rather than an _iterative method._

**Comment by tianyizheng02** [comment]
> If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

Please add the issue number to the _description_ of your PR. You'll know it worked when you see the issue appear in this section on the right-hand side of this PR:

screenshot

File: machine_learning/linear_regression.py (hunk 1 above)
Comment (tianyizheng02) [reply machine_learning/linear_regression.py:73]: This also doesn't fix the fact that the code is still calculating SSE incorrectly.

Comment (liakdimi1) [reply machine_learning/linear_regression.py:73]: can you elaborate on what do you mean by direct method?

Comment (tianyizheng02) [reply machine_learning/linear_regression.py:73]: A [direct method](https://en.wikipedia.org/wiki/Iterative_method) gives an exact solution. By contrast, an iterative method (such as gradient descent) gives increasingly accurate approximations. With an iterative method, you get _closer and closer_ to the exact solution but you may never actually reach that exact solution.

For example, let's say you want to find the roots of a quadratic polynomial $ax^2 + bx + c$. You could use Newton's method to approximate the roots, refining your solutions to be more and more accurate—this is an iterative method. Alternatively, you could just use the quadratic formula to just get the exact solutions—this is a direct method.

Comment (liakdimi1) [reply machine_learning/linear_regression.py:73]: also, explain with facts why the SSE is calculated incorrectly because the code does the predictions finds the deviation and takes the squares of the deviations, and sums the errors, which is the definition of SSE.

Comment (tianyizheng02) [reply machine_learning/linear_regression.py:73]: I already explained this with my code sample in the original issue, but sure, let's go over the math for the SSE function:
https://github.com/TheAlgorithms/Python/blob/9e08c7726dee5b18585a76e54c71922ca96c0b3a/machine_learning/linear_regression.py#L52-L64

Likes 60–62 are fine because they calculate the SSE exactly as we expect:
$$\mathrm{SSE} = \sum_{i = 0}^{N} (\mathbf{x}_i^{\top} \boldsymbol{\theta} - \mathbf{y}_i)^2$$

However, for some reason, the function then divides the SSE by `2 * len_data`. Dividing the SSE (_sum_ of squared errors) by the number of datapoints gives you the MSE (_mean_ squared error):
$$\mathrm{MSE} = \frac{1}{N} \sum_{i = 0}^{N} (\mathbf{x}_i^{\top} \boldsymbol{\theta} - \mathbf{y}_i)^2$$

Naturally, it follows that the "SSE" function is actually calculating half of the MSE because it's dividing the SSE by $2N$.

Comment (liakdimi1) [reply machine_learning/linear_regression.py:73]: i appreciate your response

**Comment by liakdimi1** [comment]
As far as the calculation of the SSE this is what i found :

when it comes to the calculation of the Sum of Squared Errors (SSE), the division by 2 * len_data is a significant factor. This division is related to the derivative of the cost function during optimization, particularly when using gradient descent. By dividing the SSE by 2 * len_data, the derivative of the cost function (SSE) with respect to the model parameters (theta) becomes simpler and results in a cleaner expression.

Additionally, it is worth noting that the cost function commonly used in linear regression is the Mean Squared Error (MSE), which is calculated as SSE divided by 2 * len_data:

MSE = SSE / (2 * len_data)

When taking the derivative of MSE with respect to the model parameters (theta), you get:

d(MSE)/d(theta) = (1/len_data) * (theta.dot(data_x.transpose()) - data_y.transpose()).dot(data_x)

Here, the division by 2 in the original SSE equation cancels out during the differentiation process, and the term 1/len_data helps scale the gradients appropriately to handle datasets of different sizes.

Although the division by 2 * len_data doesn't directly impact the optimization process, it offers a more interpretable representation of the derivative of the cost function, which can be advantageous in certain contexts.

Whether to include the division by 2 * len_data in the cost function is a matter of preference. Some implementations choose to omit this division and adjust the learning rate accordingly to compensate for the scale change. However, to ensure consistency throughout the optimization process and related calculations, it's essential to make a deliberate choice and adhere to it.

So i don't believe your speculation is correct and the implementation of the SSE stands that's from me.

**Comment by tianyizheng02** [comment]
> This division is related to the derivative of the cost function during optimization, particularly when using gradient descent. By dividing the SSE by 2 * len_data, the derivative of the cost function (SSE) with respect to the model parameters (theta) becomes simpler and results in a cleaner expression.

It's perfectly fine to scale the loss function to make the math cleaner, but again the problem here is that this scaled version is no longer the SSE. SSE has a precise mathematical definition, so why should we call the scaled version the SSE when it's clearly not? Why not just call it `scaled_sum_of_squared_errors` or `half_mean_squared_error`?

> Mean Squared Error (MSE), which is calculated as SSE divided by 2 * len_data

This is just factually wrong. MSE is defined to be SSE divided by `len_data`, _with no division by 2._ I pointed this out already in [my previous comment](https://github.com/TheAlgorithms/Python/pull/8885#discussion_r1271330609), and you can easily verify this fact with [Wikipedia](https://en.wikipedia.org/wiki/Mean_squared_error) or your search engine of choice:

$$\mathrm{MSE} = \frac{1}{n} \sum_{i = 1}^{n} (Y_i - \hat{Y}_i)^2$$

While I _have_ seen half of the MSE used in the context of ML (because, as you pointed out, it can make the math cleaner), we should definitely avoid misleading people with a different definition because this repo is for educational purposes.

Anyway, you missed the whole point of my issue. My whole point was that the output of the function `sum_of_square_error` is _not_ the SSE. This is still absolutely true, and I've already demonstrated it to you using both code and math. Plus, the fact that you've been referring to the implementation's loss function as the MSE shows that you also acknowledge that it's _not_ the SSE. If you prefer MSE (or half of the MSE) because it's more interpretable or more computationally elegant, then that's perfectly fine—I personally couldn't care less as to whether this implementation uses SSE or MSE. However, in order to not be incorrect or misleading, the function name should still be changed to reflect what it actually calculates.

**Comment by tianyizheng02** [comment]
Closing this PR since, as I've stated already, it doesn't actually resolve the linked issue
